ft-sdk/src/auth/session.rs

```diff
@@ -4,56 +4,272 @@ pub struct SessionID(pub String);
 #[cfg(feature = "auth-provider")]
 #[derive(Debug, thiserror::Error)]
 pub enum SetUserIDError {
-    #[error("session not found")]
-    SessionNotFound,
-    #[error("multiple sessions found")]
-    MultipleSessionsFound,
+    #[error("session id error: {0:?}")]
+    SessionIDError(#[from] ft_sdk::auth::SessionIDError),
     #[error("failed to query db: {0:?}")]
     DatabaseError(#[from] diesel::result::Error),
 }
 
 #[cfg(feature = "auth-provider")]
 pub fn set_user_id(
     conn: &mut ft_sdk::Connection,
-    SessionID(session_id): SessionID,
-    user_id: i64,
+    session_id: Option<SessionID>,
+    user_id: &ft_sdk::UserId,
+    session_expiration_duration: Option<chrono::Duration>,
 ) -> Result<SessionID, SetUserIDError> {
     use diesel::prelude::*;
     use ft_sdk::auth::fastn_session;
 
-    match diesel::update(fastn_session::table.filter(fastn_session::id.eq(session_id.as_str())))
-        .set(fastn_session::uid.eq(Some(user_id)))
-        .execute(conn)?
-    {
-        0 => Ok(create_with_user(conn, user_id)?),
-        1 => Ok(SessionID(session_id)),
-        _ => Err(SetUserIDError::MultipleSessionsFound),
+    let now = ft_sdk::env::now();
+
+    // Query to check if the session exists and get its expiration time
+    let session = match session_id {
+        Some(session_id) => {
+            let existing_session_expires_at = fastn_session::table
+                .select(fastn_session::expires_at.nullable())
+                .filter(fastn_session::id.eq(session_id.0.as_str()))
+                .first::<Option<chrono::DateTime<chrono::Utc>>>(conn)
+                .optional()?;
+
+            match existing_session_expires_at {
+                Some(Some(expires_at)) if expires_at < now => Some((session_id.0.clone(), true)),
+                Some(_) => Some((session_id.0.clone(), false)),
+                None => None,
+            }
+        }
+        None => match ft_sdk::auth::SessionID::from_user_id(conn, user_id) {
+            Ok(session_id) => Some((session_id.0.clone(), false)),
+            Err(ft_sdk::auth::SessionIDError::SessionExpired(session_id)) => {
+                Some((session_id, true))
+            }
+            Err(ft_sdk::auth::SessionIDError::SessionNotFound) => None,
+            Err(e) => return Err(e.into()),
+        },
+    };
+
+    match session {
+        Some((session_id, true)) => {
+            // Session is expired, delete it and create a new one
+            diesel::delete(fastn_session::table.filter(fastn_session::id.eq(session_id.as_str())))
```

Comment:
Not fond of this, we should clear up session from session table via some periodic task, and not during logout, so things can be instigated in case of problem. Deleting cookie should be enough.